### PR TITLE
[CARBONDATA-2557] [CARBONDATA-2472] [CARBONDATA-2570] Improve Carbon Reader performance on S3 and fixed datamap clear issue in reader

### PR DESCRIPTION
**Pull request by ajantha-bhat**
[CARBONDATA-2557] [CARBONDATA-2472] **Problem** : CarbonReaderBuilder.build() is slower in s3. It takes around 8 seconds to finish build()
**Solution**: S3 is slow in listFiles, open, FileExist, getCarbonFile operations. So, List down all the calls of those API in the reader flow and remove the redundant checks. 

[CARBONDATA-2570]  **Problem** : Carbon SDK Reader, second time reader instance have an issue in cluster test
**Solution**: Blocklet datamap's of first time reader is not cleared properly in the cluster. Need to change the API to clear the blocklet datamap.

so change
DataMapStoreManager.getInstance().getDefaultDataMap(queryModel.getTable()).clear();
to
DataMapStoreManager.getInstance().clearDataMaps(queryModel.getTable().getAbsoluteTableIdentifie());

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed? NA
 
 - [ ] Any backward compatibility impacted? NA
 
 - [ ] Document update required? NA

 - [ ] Testing done
        Added UT
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.  NA

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5106/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6113/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4951/

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 1]
```diff
@@ -101,6 +108,43 @@ public static CarbonReaderBuilder builder(String tablePath, String tableName) {
     return reader.readSchema();
   }
 
+  /**
+   * Read carbonindex file and return the schema
+   * @param indexFilePath complete path including index file name
+   * @return null, if the index file is not present in the path.
```
Comment (xubo245) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Can you optimize the expression? and can you add test case for returning null if the index file is not present in the path?

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 2]
```diff
@@ -101,6 +108,43 @@ public static CarbonReaderBuilder builder(String tablePath, String tableName) {
     return reader.readSchema();
   }
 
+  /**
+   * Read carbonindex file and return the schema
```
Comment (xubo245) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Can you add a empty line?

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 3]
```diff
@@ -101,6 +108,43 @@ public static CarbonReaderBuilder builder(String tablePath, String tableName) {
...
+   * @param indexFilePath complete path including index file name
+   * @return null, if the index file is not present in the path.
+   *         List<ColumnSchema> from the index file.
+   * @throws IOException
+   */
+  public static List<ColumnSchema> readSchemaInIndexFile(String indexFilePath) throws IOException {
```
Comment (xubo245) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Can you add test case for non  Transactional Table for this API?

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 4]
```diff
@@ -101,6 +108,43 @@ public static CarbonReaderBuilder builder(String tablePath, String tableName) {
...
+      dataInputStream.readFully(bytes);
+      CarbonIndexFileReader indexReader = new CarbonIndexFileReader();
+      // read from byte buffer.
+      indexReader.openThriftReader(bytes);
+      // get the index header
+      org.apache.carbondata.format.IndexHeader readIndexHeader = indexReader.readIndexHeader();
+      List<ColumnSchema> columnSchemaList = new ArrayList<ColumnSchema>();
+      List<org.apache.carbondata.format.ColumnSchema> table_columns =
+          readIndexHeader.getTable_columns();
+      for (org.apache.carbondata.format.ColumnSchema columnSchema: table_columns) {
```
Comment (xubo245) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: small code style: should add space before ':',like columnSchema : table_columns

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java (hunk 1 above)
Comment (ajantha-bhat) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Actually now it will not return null, It will throw exception. That why it is WIP. I will remove this from the method header.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java (hunk 2 above)
Comment (ajantha-bhat) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: OK

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java (hunk 3 above)
Comment (ajantha-bhat) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: This test case is redundant. As new API in this PR, just need path. It doesn't matter whether reader is transactional or not. Hence not added

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java (hunk 4 above)
Comment (ajantha-bhat) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: OK.

**Comment by xuchuanyin** [comment]
It seems the method you added is dead code since no one calls it except from test.

**Comment by ajantha-bhat** [comment]
@xuchuanyin : This is one of the exposed method in SDK reader now. External client can make use of it to get schema by passing index file path.

You can refer the other methods in this file. This is not a dead code

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6149/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4987/

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 5]
```diff
@@ -101,6 +108,45 @@ public static CarbonReaderBuilder builder(String tablePath, String tableName) {
...
+  /**
+   * Read carbonindex file and return the schema.
+   * This API is better use when index file is present,
+   * as It is faster compared to readSchemaInDataFile when carbondata file is big in size
+   *
+   * @param indexFilePath complete path including index file name
+   * @return List<ColumnSchema> from the index file.
+   * @throws IOException
+   */
+  public static List<ColumnSchema> readSchemaInIndexFile(String indexFilePath) throws IOException {
```
Comment (sounakr) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: In case we have a performance result for comparison between reading from CarbonIndexFile and CarbonDataFile, and if reading from CarbonIndexFile Shows better output, then better to read from Index File in all places, i.e. whenever internally we infer schema.

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5025/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6186/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5160/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6189/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5028/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5163/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6199/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5173/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6205/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5044/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6207/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5045/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5049/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5180/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6211/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6215/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5053/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5190/

**Comment by ajantha-bhat** [comment]
@ravipesala : please review the PR. 
With these change CarbonReaderBuilder.build() performance has improved by 50% on S3.

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 6]
```diff
@@ -996,4 +979,30 @@ public boolean canAllow(CarbonTable carbonTable, TableOperation operation) {
     }
     return indexColumn;
   }
+
+  /**
+   * update the carbon table by using the passed tableInfo
+   * @param table
```
Comment (xubo245) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: please add one blank line

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: done

**Comment by xubo245** [comment]
Have you add it in this PR?

so change
DataMapStoreManager.getInstance().getDefaultDataMap(queryModel.getTable()).clear();
to
DataMapStoreManager.getInstance().clearDataMaps(queryModel.getTable().getAbsoluteTableIdentifie());

**Comment by ajantha-bhat** [comment]
@xubo245 : Yes, It is there in PR, refer the  CarbonRecordReader.java

**Comment by xubo245** [comment]
ok, I will try It in my local machine

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6227/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5065/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5199/

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6234/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5074/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6236/

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6237/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5075/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5207/

**Comment by ajantha-bhat** [comment]
@ravipesala : please review this PR. Build is success.

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 7]
```diff
@@ -226,7 +226,7 @@ public static CarbonTable buildFromTablePath(String tableName, String tablePath,
     } else {
       // Infer the schema from the Carbondata file.
       TableInfo tableInfoInfer =
-          SchemaReader.inferSchema(AbsoluteTableIdentifier.from(tablePath, "null", "null"), false);
+          CarbonUtil.inferDummySchema(tablePath, "null", "null");
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: Change the name of the method to `buildDummyTable(tableName, tablePath)

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: OK. Changed

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5216/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6246/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5084/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5218/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5086/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6248/

**Comment by ravipesala** [comment]
LGTM
